### PR TITLE
improve getStream stability

### PR DESCRIPTION
**Pull request by rynomad**

### CODE REVIEW

File: packages/gateway-client/src/worker/index.ts [hunk 1]
```diff
@@ -17,15 +20,57 @@ import { decode, encode } from 'lob-enc';
...
 import Multiaddr from 'multiaddr';
 import * as workboxPrecaching from 'workbox-precaching';
-
 import { MessageType, ServerPeerStatus } from '../service-worker';
 
 // the workbox-precaching import includes a type definition for
 // <self dot __WB_MANIFEST>.
 // Import it even though we're not using any of the imports,
 // and mark the import as being used with this line:
+// <self dot __WB_MANIFEST>
```
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/index.ts:29]:
```suggestion
```

File: packages/gateway-client/src/worker/index.ts [hunk 2]
```diff
@@ -465,25 +553,7 @@ async function main() {
             },
         },
     });
-    // Listen for new peers
-    //let foundServer = false;
-    node.addEventListener('peer:discovery', evt => {
-        const peer = evt.detail;
-        console.log(`Found peer ${peer.id.toString()}`);
```
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/index.ts:472]: I'm thinking we should keep this console log.

**Review by JoshuaCWebDeveloper — CHANGES_REQUESTED** [review]
A few questions, minor changes, and then one significant change in the worker registration file. I can discuss the larger one with you on a call if you want.

Overall, nice solution for making the stream retry logic more efficient!

File: packages/gateway-client/src/worker/index.ts [hunk 3]
```diff
@@ -21,11 +20,58 @@ import * as workboxPrecaching from 'workbox-precaching';
...
 
 // the workbox-precaching import includes a type definition for
-// <self dot __WB_MANIFEST>.
-// Import it even though we're not using any of the imports,
-// and mark the import as being used with this line:
+// <self dot __WB_MANIFEST>
 const _ = workboxPrecaching;
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { precacheAndRoute, PrecacheEntry } from 'workbox-precaching';
```
[the remark this answers is not in the document]
Comment (JoshuaCWebDeveloper) [reply packages/gateway-client/src/worker/index.ts]: This is old I think.

File: packages/gateway-client/copy.sh [hunk 4]
```diff
@@ -1,8 +1,8 @@
 #!/bin/bash
 
 cd /usr/src
-mv ./volumes/gateway_client/assets/libp2p.bootstrap .
+mv ./volumes/gateway_client/assets/libp2p.* .
```
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/gateway-client/copy.sh:4]: ah that should've been a different PR I suppose... I'm also working on making the initial relay list more efficient.

File: packages/gateway-client/src/worker/index.ts [hunk 5]
```diff
@@ -375,8 +446,9 @@ async function main() {
         ],
         connectionManager: {
             autoDial: true, // Auto connect to discovered peers (limited by ConnectionManager minConnections)
-            minConnections: 0,
+            minConnections: 3,
```
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/gateway-client/src/worker/index.ts:533]: I think 3 is better long term: 1 direct & up to two relays.

File: packages/gateway-client/src/worker/index.ts (hunk 2 above)
Comment (rynomad) [reply packages/gateway-client/src/worker/index.ts:472]: ok, I'll add it back

File: packages/gateway-client/src/worker/index.ts [hunk 6]
```diff
@@ -591,7 +664,8 @@ self.addEventListener('fetch', function (event) {
             'track',
             'video',
             'xslt',
-        ].includes(event.request.destination)
+        ].includes(event.request.destination) ||
+        event.request.url.includes('/packs/icons/')
```
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/gateway-client/src/worker/index.ts:668]: done

File: packages/shared/service-worker/src/lib/registration.ts [hunk 7]
```diff
@@ -56,13 +56,16 @@ export function register(config?: Config) {
             });
         } else {
             // Is not localhost. Just register service worker
-            registerValidSW(swUrl, config);
+            registerValidSW(swUrl, config).catch(e => {
+                console.log('error', e);
+                registerValidSW('../service-worker.js', config);
```
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/shared/service-worker/src/lib/registration.ts:61]: ah, that got leftover from IPFS stuff, removing...

File: packages/gateway-client/copy.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/gateway-client/copy.sh:4]: to answer your question: none, but this change allows us to add some libp2p hints down the road without needing to list them explicitly

**Review by JoshuaCWebDeveloper — CHANGES_REQUESTED** [review]
Just a couple more things!

File: packages/shared/service-worker/src/lib/registration.ts [hunk 8]
```diff
@@ -56,13 +56,15 @@ export function register(config?: Config) {
             });
         } else {
             // Is not localhost. Just register service worker
-            registerValidSW(swUrl, config);
+            registerValidSW(swUrl, config).catch(e => {
+                console.log('error', e);
+            });
```
[the remark this answers is not in the document]
Comment (rynomad) [reply packages/shared/service-worker/src/lib/registration.ts]: done

File: packages/shared/service-worker/src/lib/registration.ts [hunk 9]
```diff
@@ -55,6 +55,7 @@ export function register(config?: Config) {
                 );
             });
         } else {
+            // Is not localhost. Just register service worker
```
Comment (JoshuaCWebDeveloper) [inline packages/shared/service-worker/src/lib/registration.ts:58]:
```suggestion
```

**Review by JoshuaCWebDeveloper — APPROVED** [review]

**Pull request merged**